File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokentransformer/DecryptTokenTransformer.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokentransformer;

import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Transforms the token using AES-256 symmetric decryption.
 * 
 * @see <a href=https://datatracker.ietf.org/doc/html/rfc3826>AES</a>
 */
public class DecryptTokenTransformer implements TokenTransformer {
    private static final long serialVersionUID = 1L;

    private static final Logger logger = LoggerFactory.getLogger(DecryptTokenTransformer.class);

    private final SecretKeySpec secretKey;

    /**
     * Initializes the underlying cipher (AES) with the decryption secret.
     * 
     * @param encryptionKey the encryption key. The key must be 32 characters long.
     * 
     * @throws java.security.InvalidKeyException                invalid encryption
     *                                                          key.
     * @throws java.security.InvalidAlgorithmParameterException invalid encryption
     *                                                          algorithm
     *                                                          parameters.
     */
    public DecryptTokenTransformer(String encryptionKey)
            throws InvalidKeyException, InvalidAlgorithmParameterException {
        if (encryptionKey.length() != EncryptionConstants.KEY_BYTE_LENGTH) {
            logger.error("Invalid Argument. Key must be {} characters long", EncryptionConstants.KEY_BYTE_LENGTH);
            throw new InvalidKeyException(
                    String.format("Key must be %s characters long", EncryptionConstants.KEY_BYTE_LENGTH));
        }

        this.secretKey = new SecretKeySpec(encryptionKey.getBytes(StandardCharsets.UTF_8), EncryptionConstants.AES);
    }

    /**
     * Decryption token transformer.
     * <p>
     * Decrypts the token using AES-256 symmetric decryption algorithm.
     *
     * @return the decrypted token string.
     * @param token the encrypted token in base64 format.
     * @throws java.lang.IllegalStateException        the underlying cipher
     *                                                is in a wrong state.
     * @throws javax.crypto.IllegalBlockSizeException if this cipher is a block
     *                                                cipher,
     *                                                no padding has been requested
     *                                                (only in encryption mode), and
     *                                                the total
     *                                                input length of the data
     *                                                processed by this cipher is
     *                                                not a multiple of
     *                                                block size; or if this
     *                                                encryption algorithm is unable
     *                                                to
     *                                                process the input data
     *                                                provided.
     * @throws javax.crypto.BadPaddingException       invalid padding size.
     * @throws InvalidAlgorithmParameterException     invalid encryption
     * @throws InvalidKeyException                    invalid encryption key.
     * @throws java.security.NoSuchAlgorithmException invalid encryption
     *                                                algorithm/mode.
     * @throws javax.crypto.NoSuchPaddingException    invalid encryption
     *                                                algorithm padding.
     */
    @Override
    public String transform(String token)
            throws IllegalStateException, IllegalBlockSizeException, BadPaddingException, InvalidKeyException,
            InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchPaddingException {

        // Decode the base64-encoded token
        byte[] messageBytes = Base64.getDecoder().decode(token);

        // Extract IV and ciphertext
        byte[] ivBytes = new byte[EncryptionConstants.IV_SIZE];
        byte[] cipherBytes = new byte[messageBytes.length - EncryptionConstants.IV_SIZE];

        System.arraycopy(messageBytes, 0, ivBytes, 0, EncryptionConstants.IV_SIZE);
        System.arraycopy(messageBytes, EncryptionConstants.IV_SIZE, cipherBytes, 0, cipherBytes.length);

        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(EncryptionConstants.TAG_LENGTH_BITS, ivBytes);

        // Initialize AES cipher in GCM mode with no padding for decryption
        Cipher cipher = Cipher.getInstance(EncryptionConstants.ENCRYPTION_ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, this.secretKey, gcmParameterSpec);

        byte[] decryptedBytes = cipher.doFinal(cipherBytes);

        return new String(decryptedBytes, StandardCharsets.UTF_8);
    }
}

File: lib/python/opentoken/src/main/opentoken/tokentransformer/decrypt_token_transformer.py
"""
Copyright (c) Truveta. All rights reserved.
"""
import base64
import logging
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from opentoken.tokentransformer.token_transformer import TokenTransformer
from opentoken.tokentransformer.encryption_constants import EncryptionConstants


logger = logging.getLogger(__name__)


class DecryptTokenTransformer(TokenTransformer):
    """AES-256 GCM token decryption transformer.

    Parity with Java `DecryptTokenTransformer`:
    - Java prepends IV to (ciphertext || tag) and GCM consumes combined remainder.
    - Python splits IV, ciphertext, and tag explicitly before constructing GCM mode.
    - Both expect 12-byte IV and 16-byte tag; key length enforced at init.
    """

    def __init__(self, encryption_key: str):
        """
        Initializes the underlying cipher (AES) with the decryption secret.

        Args:
            encryption_key: The encryption key. The key must be 32 characters long.

        Raises:
            ValueError: If the encryption key is not 32 characters long.
        """
        if len(encryption_key) != EncryptionConstants.KEY_BYTE_LENGTH:
            logger.error(f"Invalid Argument. Key must be {EncryptionConstants.KEY_BYTE_LENGTH} characters long")
            raise ValueError(f"Key must be {EncryptionConstants.KEY_BYTE_LENGTH} characters long")

        self.encryption_key = encryption_key.encode('utf-8')

    def transform(self, token: str) -> str:
        """Decrypt a base64 encoded token produced by the encrypt transformer.

        Args:
            token: Base64 string containing IV || ciphertext || tag.

        Returns:
            Decrypted UTF-8 token string.

        Raises:
            Exception: Propagates cryptographic failures (logged here first).
        """
        try:
            # Decode the base64-encoded token
            message_bytes = base64.b64decode(token)

            # Extract IV, encrypted data, and tag
            iv_bytes = message_bytes[:EncryptionConstants.IV_SIZE]
            ciphertext_and_tag = message_bytes[EncryptionConstants.IV_SIZE:]
            ciphertext = ciphertext_and_tag[:-EncryptionConstants.TAG_LENGTH_BYTES]
            tag = ciphertext_and_tag[-EncryptionConstants.TAG_LENGTH_BYTES:]

            # Create cipher for decryption
            cipher = Cipher(
                algorithms.AES(self.encryption_key),
                modes.GCM(iv_bytes, tag),
                backend=default_backend()
            )

            # Decrypt the token
            decryptor = cipher.decryptor()
            decrypted_bytes = decryptor.update(ciphertext) + decryptor.finalize()

            return decrypted_bytes.decode('utf-8')

        except Exception as e:
            logger.error("Error during token decryption", exc_info=e)
            raise
